agent/taskresource/volume.go

```diff
@@ -0,0 +1,146 @@
+func NewVolumeResource(name string, 
+	mountPoint string, 
+	driver string, 
+	driverOptions map[string]string, 
+	labels map[string]string) *VolumeResource {
+
+	vol := &VolumeResource{}
+	vol.Name = name
+	vol.Mountpoint = mountPoint
+	vol.Driver = driver
+	vol.DriverOptions = driverOptions
+	vol.Labels = labels
+	return vol
+}
+
+// SetDesiredStatus safely sets the desired status of the resource
+func (vol *VolumeResource) SetDesiredStatus(status VolumeStatus) {
+	vol.lock.Lock()
+	defer vol.lock.Unlock()
+
+	vol.desiredStatusUnsafe = status
+}
+
+// GetDesiredStatus safely returns the desired status of the task
+func (vol *VolumeResource) GetDesiredStatus() VolumeStatus {
+	vol.lock.RLock()
+	defer vol.lock.RUnlock()
+
+	return vol.desiredStatusUnsafe
+}
+
+// SetKnownStatus safely sets the currently known status of the resource
+func (vol *VolumeResource) SetKnownStatus(status VolumeStatus) {
+	vol.lock.Lock()
+	defer vol.lock.Unlock()
+
+	vol.knownStatusUnsafe = status
+}
+
+// GetKnownStatus safely returns the currently known status of the task
+func (vol *VolumeResource) GetKnownStatus() VolumeStatus {
+	vol.lock.RLock()
+	defer vol.lock.RUnlock()
+
+	return vol.knownStatusUnsafe
+}
+
+// SetCreatedAt sets the timestamp for resource's creation time
+func (vol *VolumeResource) SetCreatedAt(createdAt time.Time) {
+	if createdAt.IsZero() {
+		return
+	}
+	vol.lock.Lock()
+	defer vol.lock.Unlock()
+
+	vol.createdAt = createdAt
+}
+
+// GetCreatedAt sets the timestamp for resource's creation time
+func (vol *VolumeResource) GetCreatedAt() time.Time {
+	vol.lock.RLock()
+	defer vol.lock.RUnlock()
+
+	return vol.createdAt
+}
+
+// volumeResourceJSON duplicates VolumeResource fields, only for marshalling and unmarshalling purposes
+type volumeResourceJSON struct {
+	Name          string            `json:"Name"`
+	Mountpoint    string            `json:"MountPoint"`
+	Driver        string            `json:"Driver"`
+	DriverOptions map[string]string `json:"DriverOptions"`
+	Labels        map[string]string `json:"Labels"`
+
+	DesiredStatusUnsafe *VolumeStatus `json:"DesiredStatus"`
+	KnownStatusUnsafe   *VolumeStatus `json:"KnownStatus"`
+}
+
+// MarshalJSON marshals VolumeResource object using duplicate struct VolumeResourceJSON
+func (vol *VolumeResource) MarshalJSON() ([]byte, error) {
+	return json.Marshal(volumeResourceJSON{
+		vol.Name,
+		vol.Mountpoint,
+		vol.Driver,
+		vol.DriverOptions,
+		vol.Labels,
+		&vol.desiredStatusUnsafe,
+		&vol.knownStatusUnsafe,
+	})
+}
+
+// UnmarshalJSON unmarshals VolumeResource object using duplicate struct VolumeResourceJSON
+func (vol *VolumeResource) UnmarshalJSON(b []byte) error {
+	temp := &volumeResourceJSON{}
+
+	if err := json.Unmarshal(b, &temp); err != nil {
+		return err
+	}
+
+	vol.Name = temp.Name
+	vol.Mountpoint = temp.Mountpoint
+	vol.Driver = temp.Driver
+	vol.DriverOptions = temp.DriverOptions
+	vol.Labels = temp.Labels
+	vol.desiredStatusUnsafe = *temp.DesiredStatusUnsafe
```

Comment:
there's a potential NPE here. Can you please add a `nil` check here and at the next line?